src/sources.rs

```diff
@@ -119,3 +121,47 @@ impl<A, St, F> Iterator for Unfold<St, F>
         (0, None)
     }
 }
+
+/// An iterator that infinitely applies function to state and yields results.
+///
+/// This `struct` is created by the [`iterate()`] function. See its documentation for more.
+///
+/// [`iterate()`]: ../fn.iterate.html
+#[derive(Clone)]
+pub struct Iterate<St, F> {
+    state: St,
+    f: F,
+}
+
+impl<St, F> Iterator for Iterate<St, F>
+    where F: FnMut(&St) -> St
+{
+    type Item = St;
+
+    #[inline]
+    fn next(&mut self) -> Option<Self::Item> {
+        let next_state = (self.f)(&self.state);
+        Some(mem::replace(&mut self.state, next_state))
+    }
+
+    #[inline]
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        (usize::max_value(), None)
+    }
+}
+
+/// Creates a new iterator that infinitely applies function to state and yields results.
```

Comment:
The doc is actually missing the detail here that the "state" is updated with the new result each time. I think maybe it's best not to talk of it as "state" at all, but as applying a function on a value.